pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,539 @@
+		let mut message = None;
+		let mut votes = 0;
+
+		for router_id in &inbound_processing_info.router_ids {
+			match PendingInboundEntries::<T>::get(message_proof, router_id) {
+				// We expected one InboundEntry for each router, if that's not the case,
+				// we can return.
+				None => return Ok(()),
+				Some(stored_inbound_entry) => match stored_inbound_entry {
+					InboundEntry::Message(message_entry) => message = Some(message_entry.message),
+					InboundEntry::Proof(proof_entry) => {
+						if proof_entry
+							.has_valid_vote_for_session(inbound_processing_info.current_session_id)
+						{
+							votes.ensure_add_assign(1)?;
+						}
+					}
+				},
+			};
+		}
+
+		if votes < inbound_processing_info.expected_proof_count_per_message {
+			return Ok(());
+		}
+
+		match message {
+			Some(msg) => {
+				Self::execute_post_voting_dispatch(&inbound_processing_info, message_proof)?;
+
+				T::InboundMessageHandler::handle(
+					inbound_processing_info.domain_address.clone(),
+					msg,
+				)
+			}
+			None => Ok(()),
+		}
+	}
+
+	/// Decreases the counts for inbound entries and removes them if the
+	/// counts reach 0.
+	pub(crate) fn execute_post_voting_dispatch(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message_proof: Proof,
+	) -> DispatchResult {
+		for router_id in &inbound_processing_info.router_ids {
+			PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+				match storage_entry {
+					None => {
+						Err::<(), DispatchError>(Error::<T>::PendingInboundEntryNotFound.into())
+					}
+					Some(stored_inbound_entry) => {
+						let post_dispatch_entry = InboundEntry::create_post_voting_entry(
+							&stored_inbound_entry,
+							&inbound_processing_info,
+						)?;
+
+						*storage_entry = post_dispatch_entry;
+
+						Ok(())
+					}
+				}
+			})?;
+		}
+
+		Ok(())
+	}
+
+	/// Iterates over a batch of messages and checks if the requirements for
+	/// processing each message are met.
+	pub(crate) fn process_inbound_message(
+		domain_address: DomainAddress,
+		message: T::Message,
+		router_id: T::RouterId,
+	) -> (DispatchResult, Weight) {
+		let weight = LP_DEFENSIVE_WEIGHT;
+
+		let inbound_processing_info = match domain_address.clone().try_into() {
+			Ok(i) => i,
+			Err(e) => return (Err(e), weight),
+		};
+
+		let mut count = 0;
+
+		for submessage in message.submessages() {
+			if let Err(e) = count.ensure_add_assign(1) {
+				return (Err(e.into()), weight.saturating_mul(count));
+			}
+
+			let message_proof = Self::get_message_proof(message.clone());
+
+			if let Err(e) = Self::validate_and_upsert_pending_entries(
+				&inbound_processing_info,
+				submessage.clone(),
+				message_proof,
+				router_id.clone(),
+			) {
+				return (Err(e), weight.saturating_mul(count));
+			}
+
+			match Self::execute_if_requirements_are_met(&inbound_processing_info, message_proof) {
+				Err(e) => return (Err(e), weight.saturating_mul(count)),
+				Ok(_) => continue,
+			}
```

Comment:
To me this seems like a 2 part kind of process:
1 - get the initial entry and update counts;
2 - check the counts, execute if needed and update counts again.

Reply:
But the above without the second one can not exist, right? Because it will leave the state in a wrong state. Not a strong opinion about this point! Too much nitpick from my side

Reply:
I stil changed some things in there, mainly dropped `validate_and_upsert_pending_entries`, most of the logic is now in `process_inbound_message`.